### ISSUE TITLE
TPS overly drops on my map.

### ISSUE THREAD

**Issue by BlackLion1411** [Stardust-Labs-MC/Terralith#185]
### What Minecraft version(s) does this apply to?

1.21 - 1.21.1

### Project version

v2.5.7

### Mod or datapack version?

Fabric/Quilt Mod

### Describe the bug

I see on several biomes my TPS drop by A LOT. And makes the game quite unplayable. Overall performance goes great (FPS and chunk loading goes smooth), but the TPS are the ones giving me trouble. There are no overflow of entities/dropped items rendering, it just happens when exploring certain biomes, which occurs quite regularly (And it's not other mods, bc as soon as I go to another dimension, TPS go up, I strongly believe it is because the generation of certain blocks in the map, or an unstable color in the generation of leaves of the trees). I get around 11-6 TPS average on these zones. And I can't see what might be the problem with the map loading. There's no lagometer in my modpack version to see what is stressing the chunk loading so much. It only becomes normal at 8* chunks but that even might be not enough. And I play normally at 16 chunks in order to be stable at my base, bc at 32 it's just completely unplayable on every place. And I had the v2.5.6 for the moment until I saw I could update, I thought it might solve my problem, but it didn't.

I'm also using "Fabulously Optimized" modpack. I don't think it's the cause, but let me know if there's any performance mod that causes huge trouble with Terralith. (I have this mod in order to achieve stable V-Sync 144FPS with complementary shaders) I'd be grateful if you could help me. This mod is beautiful and I don't wan't to play without it.

![TPS](https://github.com/user-attachments/assets/25853d8d-3e8d-4e8c-9efb-fd4d43be25a6)

**Comment by catter1** [comment]
This is odd, as it's not really possible for Terralith to be causing TPS drops in this scenario. The only time Terralith can realistically cause TPS drops:
- Generating new chunks for the first time
- Loading existing chunks (but way less of a performance hit than the initial generation)
- Lots of water updates (aka when aquifers spill in caves or waterfalls generate, but again this only happens the first time a chunk is generated)

So tl;dr, once a chunk is generated, Terralith has zero impact on performance for that chunk.

For generating a lot of new chunks, yes, Terralith *can* be a source of lag. For that, we always highly recommend people to pregen their world using the [Chunky mod](https://modrinth.com/plugin/chunky). But based on the wording of your issue, it is implied to me that the issue persists in already loaded chunks.

## Response
To address some of your ideas:
> the generation of certain blocks in the map

No, as we only use natural vanilla blocks, which would not be laggier than any other block in any other biome

> an unstable color in the generation of leaves of the trees

Not possible for two reasons:
- If color was theoretically an issue, it would cause FPS lag, not TPS
- We use the *exact* same system vanilla uses to color their leaves. For example, Jungle has bright green leaves and Swamp dark green. It's the exact same when we choose red or purple or whatever color leaves, and would mean zero performance difference

## Suggestion
While there is no "lag-o-meter" per se, you could try using [Spark](https://modrinth.com/mod/spark) to see *what exactly* is causing the performance hit. Is it entities? Is it some weird worldgen process? Is it perhaps a mod performing incorrectly? Now, spark isn't able to tell you if a mod is causing an issue, but where it highlights the TPS drops would give a nudge in the right direction.

It's simply extremely unlikely (I don't like saying "impossible") for Terralith to be the cause of your performance issue. I would recommend messing around with disabling/enabling some of your mods to see if any of them help with performance.

**Comment by BlackLion1411** [comment]
So I just solved it... Turns out it was a mod called "Bigger stack sizes" that was having conflict with other mod called "StackItem64" in a chest full of eggs stacked to 64 near this spot (On the left of the last image, somewhat under the cobble stairs)

But I SWEAR to you, I had this problem of TPS *long before installing these QOL mods, *long before making the egg farm. Now I stuck up with StackItem64 as the BSS mod was the conflictive one so I deleted it.

I also believe Fabulously Optimized modpack was causing trouble so I removed the files of the modpack and reinstalled the last release shown in modrinth, because I used the Spark mod as you suggested me to see what was causing the issues, and DynamicFPS within Modernfix were the ones consuming the most resources in a proccess called "Unsafe.Park" as shown in the following image. Which was, according to ChatGPT, a faulty optimization problem with a task overflow. So the reinstallation it's a "Just in case".

I also installed Chunky as another "Just In case" while testing for the problematic mod, and none of the other mods made conflict with performance. And with this I also hope new world generation may be more stable while exploring new chunks.

So It was not anything related to Terralith, nor other important mods for me, and I might've been confusing the new generation world TPS dropout (That naturally happened but would be stable later) with the bug happening in my base's 16x16 chunks surroundings with the egg farm. I still think that was weird af, because as I said earlier, my problem was happening before that mod and before that farm, in a lot of different biomes (Already explored).

So TL:DR: It's solved (I hope, but I doubt I'll have more trouble as I checked on the rest of the base surroundings and TPS are stable) the problem was a fricking egg farm. Now I can play smoothly at 32 chunks render distance without trouble and it looks magnificent. Thanks for the Spark and Chunky suggestion, it was very useful for me.

Here's the Spark data, which was related to DynamicFPS and Modernfix:

![TPS](https://github.com/user-attachments/assets/d7225a69-4699-4715-bbdc-231791dde70e)
